lib/srv/desktop/rdp/rdpclient/src/network_client.rs

```diff
@@ -0,0 +1,154 @@
+use ironrdp_connector::sspi::network_client::NetworkProtocol;
+use ironrdp_connector::{custom_err, ConnectorResult};
+use ironrdp_tokio::AsyncNetworkClient;
+use reqwest::Client;
+use sspi::{Error, ErrorKind};
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
+use tokio::net::{TcpStream, UdpSocket};
+use url::Url;
+
+pub(crate) struct ReqwestNetworkClient {
+    client: Option<Client>,
+}
+
+impl AsyncNetworkClient for ReqwestNetworkClient {
+    fn send<'a>(
+        &'a mut self,
+        request: &'a NetworkRequest,
+    ) -> Pin<Box<dyn Future<Output = ConnectorResult<Vec<u8>>> + 'a>> {
+        Box::pin(async move {
+            match &request.protocol {
+                NetworkProtocol::Tcp => self.send_tcp(&request.url, &request.data).await,
+                NetworkProtocol::Udp => self.send_udp(&request.url, &request.data).await,
+                NetworkProtocol::Http | NetworkProtocol::Https => {
+                    self.send_http(&request.url, &request.data).await
+                }
+            }
+        })
+    }
+}
+
+impl ReqwestNetworkClient {
+    pub(crate) fn new() -> Self {
+        Self { client: None }
+    }
+}
+
+// 65535 bytes: maximum allowed token len in Windows
+// See note in https://learn.microsoft.com/en-us/troubleshoot/windows-server/group-policy/group-policy-add-maxtokensize-registry-entry
+const MAX_TOKEN_LENGTH: usize = 65536;
+const DEFAULT_KERBEROS_PORT: u16 = 88;
+
+impl ReqwestNetworkClient {
+    async fn send_tcp(&self, url: &Url, data: &[u8]) -> ConnectorResult<Vec<u8>> {
+        let addr = format!(
+            "{}:{}",
+            url.host_str().unwrap_or_default(),
+            url.port().unwrap_or(88)
+        );
+
+        let mut stream = TcpStream::connect(addr)
+            .await
+            .map_err(|e| Error::new(ErrorKind::NoAuthenticatingAuthority, format!("{:?}", e)))
+            .map_err(|e| custom_err!("failed to connect to KDC over TCP", e))?;
+
+        stream
+            .write(data)
+            .await
+            .map_err(|e| Error::new(ErrorKind::NoAuthenticatingAuthority, format!("{:?}", e)))
+            .map_err(|e| custom_err!("failed to send KDC request over TCP", e))?;
+
+        let len = stream
+            .read_u32()
+            .await
+            .map_err(|e| Error::new(ErrorKind::NoAuthenticatingAuthority, format!("{:?}", e)))
+            .map_err(|e| custom_err!("failed to read KDC response length over TCP", e))?;
+
+        let mut buf = vec![0; len as usize + 4];
+        buf[0..4].copy_from_slice(&(len.to_be_bytes()));
+
+        stream
+            .read_exact(&mut buf[4..])
+            .await
+            .map_err(|e| Error::new(ErrorKind::NoAuthenticatingAuthority, format!("{:?}", e)))
+            .map_err(|e| custom_err!("failed to send KDC response over TCP", e))?;
+
+        Ok(buf)
+    }
+
+    async fn send_udp(&self, url: &Url, data: &[u8]) -> ConnectorResult<Vec<u8>> {
+        let udp_socket = UdpSocket::bind((IpAddr::V4(Ipv4Addr::LOCALHOST), 0))
```

Comment:
Is `LOCALHOST` okay here?

Reply:
I simplify the client, now we error out for HTTP, HTTPS and UDP